middlewares/mw/middleware.go

```diff
@@ -0,0 +1,40 @@
+package mw
+
+import "net/http"
+
+// Middleware types should embed BasicMiddleware and implement ServeHTTP and Close as required.
+type Middleware interface {
+	http.Handler
+	Next() http.Handler
+	Close()
+}
+
+// BasicMiddleware should be embedded in Middleware types. These implement ServeHTTP, and must
+// be careful to call the next middleware (unless explicitly require not to).
+//
+// Close should also be overridden if required.
+type BasicMiddleware struct {
+	next http.Handler
+}
+
+// NewMiddleware creates a new BasicMiddleware.
+func NewMiddleware(next http.Handler) BasicMiddleware {
+	return BasicMiddleware{next}
+}
+
+// Next returbs the next middleware in the chain.
+func (m *BasicMiddleware) Next() http.Handler {
+	return m.next
+}
+
+// Close closes the next middleware in the chain but does not do anything to this one.
+// This method should be overridden if specialised behaviour is needed.
+// Note that any errors should be logged; they are not propagated back.
```

Comment:
The middlewares are strung together in a chain.  The default BasicMiddleware as nothing to close; all it needs to do is call close on the next middleware in the chain in case it does have something to close.  Any middleware that does have something to close provides it's own Close function and must also call Close on the next in the chain.